evidence/reactor.go

```diff
@@ -21,233 +46,346 @@ const (
+		// a new done channel so we can explicitly close the goroutine if the peer
+		// is later removed, we increment the waitgroup so the reactor can stop
+		// safely, and finally start the goroutine to broadcast evidence to that peer.
+		_, ok := r.peerRoutines[peerUpdate.PeerID.String()]
+		if !ok {
+			r.peerRoutines[peerUpdate.PeerID.String()] = make(chan struct{})
+			r.peerWG.Add(1)
+			go r.broadcastEvidenceLoop(peerUpdate.PeerID, r.peerRoutines[peerUpdate.PeerID.String()])
+		}
+
+	case p2p.PeerStatusDown, p2p.PeerStatusRemoved, p2p.PeerStatusBanned:
+		// Check if we've started an evidence broadcasting goroutine for this peer.
+		// If we have, we signal to terminate the goroutine via the channel's closure.
+		// This will internally decrement the peer waitgroup and remove the peer
+		// from the map of peer evidence broadcasting goroutines.
+		doneCh, ok := r.peerRoutines[peerUpdate.PeerID.String()]
+		if ok {
+			close(doneCh)
+		}
+	}
+
+	return err
 }
 
-// Modeled after the mempool routine.
-// - Evidence accumulates in a clist.
-// - Each peer has a routine that iterates through the clist,
-// sending available evidence to the peer.
-// - If we're waiting for new evidence and the list is not empty,
-// start iterating from the beginning again.
-func (evR *Reactor) broadcastEvidenceRoutine(peer p2p.Peer) {
+// processPeerUpdates initiates a blocking process where we listen for and handle
+// PeerUpdate messages. When the reactor is stopped, we will catch the signal and
+// close the p2p PeerUpdatesCh gracefully.
+func (r *Reactor) processPeerUpdates() {
+	defer r.peerUpdates.Close()
+
+	for {
+		select {
+		case peerUpdate := <-r.peerUpdates.Updates():
+			if err := r.processPeerUpdate(peerUpdate); err != nil {
+				r.Logger.Error(
+					"failed to process peer update",
+					"peer", peerUpdate.PeerID.String(),
+					"status", peerUpdate.Status,
+					"err", err,
+				)
+			}
+
+		case <-r.closeCh:
+			r.Logger.Debug("stopped listening on peer updates channel; closing...")
+			return
+		}
+	}
+}
+
+// broadcastEvidenceLoop starts a blocking process that continuously reads pieces
+// of evidence off of a linked-list and sends the evidence in a p2p Envelope to
+// the given peer by ID. This should be invoked in a goroutine per unique peer
+// ID via an appropriate PeerUpdate. The goroutine can be signaled to gracefully
+// exit by either explicitly closing the provided doneCh or by the reactor
+// signaling to stop.
+//
+// TODO: This should be refactored so that we do not blindly gossip evidence
+// that the peer has already received or may not be ready for.
+//
+// REF: https://github.com/tendermint/tendermint/issues/4727
+func (r *Reactor) broadcastEvidenceLoop(peerID p2p.PeerID, doneCh <-chan struct{}) {
 	var next *clist.CElement
+
+	defer func() {
+		if e := recover(); e != nil {
+			r.Logger.Error("recovering from processing peer update panic", "err", e)
+			r.removePeerRoutine(peerID)
```

Comment:
We may want to remove the peer routine here in all cases, since we want to guarantee that it's cleaned up when the function returns in all cases. Then we can just inline the logic as well, since we won't have to call it everywhere.